lib/Source/AutoHog/AutoHogImporter.cpp

```diff
@@ -0,0 +1,530 @@
+    const char *portDirection = port["direction"].GetString();
+
+    Port ioPort;
+    ioPort.portName = std::string(portName);
+
+    if (strcmp(portDirection, "input") == 0) {
+      ioPort.type = PortType::INPUT;
+    } else if (strcmp(portDirection, "output") == 0) {
+      ioPort.type = PortType::OUTPUT;
+    } else {
+      assert(false && "Unknown port direction, expected 'input or 'output'");
+    }
+
+    ports[ioPort.portName] = ioPort;
+  }
+
+  // Second iteration to check for cell/vs IO port connection.
+  // Confusing because "bits" field can be referring to either an io port
+  // or a cell name.
+  for (rapidjson::Value::ConstValueIterator itr = docPorts.Begin();
+       itr != docPorts.End(); ++itr) {
+    const rapidjson::Value &port = *itr;
+    const char *portName = port["name"].GetString();
+    Port ioPort = ports[portName];
+    LLVM_DEBUG(llvm::dbgs() << "Checking if IO port " << portName
+                            << " has a direct IO connection\n");
+
+    if (ioPort.type == PortType::OUTPUT) {
+      if (port.HasMember("bits")) {
+        auto bits = port["bits"].GetArray();
+        assert(bits.Size() == 1 && "Expected 'bits' to be an array of size 1");
+        auto *connectedPortName = bits[0].GetString();
+        if (ports.contains(connectedPortName) &&
+            ports[connectedPortName].type == PortType::INPUT) {
+          LLVM_DEBUG(llvm::dbgs() << "IO port " << portName
+                                  << " has a direct IO connection to "
+                                  << connectedPortName << "\n");
+          ioPort.connectedInputPort = connectedPortName;
+        }
+      }
+    }
+
+    ports[ioPort.portName] = ioPort;
+  }
+}
+
+/// Parse the connections field of a cell object in the JSON document.
+/// - Populates the connections map, mapping the local port name to the
+/// connected
+///   Port object (an IO port or another cell).
+/// - Populates the outputPortToConnectedCellPort map, mapping an output ioPort
+///   to the corresponding local CELL type Port of the current cell.
+void parseCellConnections(const char *cellName,
+                          const rapidjson::Value &cellData,
+                          const StringMap<Port> &ioPorts,
+                          StringMap<Port> &connections,
+                          StringMap<Port> &outputPortToConnectedCellPort) {
+  LLVM_DEBUG(llvm::dbgs() << "Parsing cell connections\n");
+  auto &jsonConnections = cellData["connections"];
+  assert(jsonConnections.IsObject() &&
+         "Expected 'connections' to be an object");
+  for (rapidjson::Value::ConstMemberIterator itr =
+           jsonConnections.MemberBegin();
+       itr != jsonConnections.MemberEnd(); ++itr) {
+    const char *localPortName = itr->name.GetString();
+    const rapidjson::Value &portData = itr->value;
+    assert(portData.IsObject() &&
+           "Expected 'connections[...]' to be an object");
+    const char *connectedPortName = portData["port"].GetString();
```

Comment:
I don't really understand why this value is always needed - in toposort.json the AND cell's port `$0$` connection contains both `cell` and `port` - 

```
        "$0$": {
          "cell": "input_cell",
          "port": "$2$"
        },
```

Shouldn't it either be connected to a port or a cell? Why does this field have both? It's also confusing because that port name is the same name as a local port name...

Reply:
I feel the names are also confusing, but the way I understand it is that `port` refers to one of:

- Global circuit input/output (when `cell` is not present)
- Local cell input/output (when `cell` is present)

The problem is that `$2$` is used many times in the example JSON, in each cell that has at least three inputs. So the way to disambiguate is either to give everything unique names, or add the name of the cell as a sort of namespace for the ports local to that cell.